File: prep/implementation/patterns_reference_imeplementation/sliding_window/AverageOfSubarraysOfSizeK.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

// O(n) / O(1)
public class AverageOfSubarraysOfSizeK {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int k = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        double[] res = new double[n - k + 1];
        int sum = 0, windowStart = 0;

        for (int windowEnd = 0; windowEnd < n; windowEnd++) {
          sum += a[windowEnd];

          if (windowEnd >= k - 1) {
            res[windowStart] = (double)sum / k;
            sum -= a[windowStart];
            windowStart++;
          }
        }

        for (double r : res) {
          System.out.printf("%.1f ", r);
        }
        System.out.println();
      }
    }
  }
}

File: prep/implementation/patterns_reference_imeplementation/sliding_window/average_of_subarrays_of_size_k.rs
use std::io;

fn get_pair() -> (i32, i32) {
    let mut input = String::new();

    io::stdin().read_line(&mut input).unwrap();
    let v = input
        .trim()
        .split_whitespace()
        .map(|d| d.parse::<i32>().unwrap())
        .collect::<Vec<_>>();

    (v[0].clone(), v[1].clone())
}

fn get_nums() -> Vec<i32> {
    let mut input = String::new();

    io::stdin().read_line(&mut input).unwrap();
    input
        .trim()
        .split_whitespace()
        .map(|d| d.parse::<i32>().unwrap())
        .collect::<Vec<_>>()
}

fn get_string() -> String {
    let mut input = String::new();

    io::stdin().read_line(&mut input).unwrap();
    input.trim().to_owned()
}

fn get_num() -> i32 {
    let mut input = String::new();

    io::stdin().read_line(&mut input).unwrap();
    input.trim().parse::<i32>().unwrap()
}

fn main() {
    let (n, k) = get_pair();
    let a = get_nums();

    let mut sum = 0.0;
    let mut window_start = 0usize;
    let mut window_end = 0usize;
    let mut res = Vec::with_capacity(n as usize - k as usize + 1);

    while window_end < n as usize {
        sum += a[window_end] as f64;

        if window_end >= k as usize - 1 {
            res.push(sum / (k as f64));
            sum -= a[window_start] as f64;
            window_start += 1;
        }
        window_end += 1;
    }

    for r in res {
        print!("{:.2} ", r);
    }
    println!();
}